### PR TITLE
Fix flush interval

### LINKED ISSUE #337: consumer auto commit flush interval reset by processed messages
### Is your feature request related to a problem? Please describe.

The stream consumer has an auto commit strategy that can be enabled, which defaults to commit the offset every 10.000 messages or 5 seconds (flush interval). The timer for the flush interval is however reset every time a message is processed. The problem is that if messages are constantly received before the flush interval, then the flush interval is never triggered. 

If we have one message being processed every second, then the default flush interval of 5 seconds is never hit and the offset is only committed when closing the consumer gracefully or once 10.000 messages have been processed. If the connection to the broker is lost or the consumer was not gracefully closed, then this may result in a lot of messages being replayed, up to 10.000 by default.

### Describe the solution you'd like

The offset should be committed at the flush interval regardless of incoming messages.

### Describe alternatives you've considered

For low volume systems the messageCountBeforeStorage can be drastically lowered, however under most unfortunate circumstances the offset may not be committed for the duration of `messageCountBeforeStorage * flushInterval`. If I lower messageCountBeforeStorage to 100 and keep flushInterval at 5 seconds, then the offset may not be committed for up to 500 seconds. This is too long and bad for performance to commit the offset so often.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Darthmineboy**
Fixes auto commit at flush interval not being triggered due to constant stream of messages which resets the flush interval.

I was having a bit of an issue with CI that it says `There were failures detected in the following suites: stream ./pkg/stream` but I was unable to see the failing test. I upgraded ginkgo because I thought it may be related to https://github.com/onsi/ginkgo/issues/973 where test summary is not shown when just 1 test fails. 

Turns out that this failure was caused by a data race that was logged as warning but it doesn't mark a specific test as failed. The data race is now fixed. If you want I can revert the ginkgo upgrade.

fixes #337

### CODE REVIEW

**Comment by Gsantomaggio** [comment]
Thank you. I put @hiimjako in the cc to check the PR.

File: pkg/stream/consumer_test.go [hunk 1]
```diff
@@ -282,6 +282,36 @@ var _ = Describe("Streaming Consumers", func() {
...
+			time.Sleep(time.Millisecond * 100)
+
+			if consumer.GetLastStoredOffset() > 0 {
+				break
+			}
+		}
+
+		Expect(messagesReceived > 5 && messagesReceived < int32(maxMessages)).To(BeTrueBecause("%d messages received", messagesReceived))
+		Expect(producer.Close()).NotTo(HaveOccurred())
+		Expect(consumer.Close()).NotTo(HaveOccurred())
```
Comment (hiimjako) [inline pkg/stream/consumer_test.go:312]:
```suggestion
```

**Comment by hiimjako** [comment]
Looks great to me!

I was also thinking:
1. If the lock in `cacheStoreOffset` is needed, because it doesn't look like there are concurrent calls to the function.
1. If it makes sense to move the reset of `messageCountBeforeStorage` into `cacheStoreOffset`, along with `lastAutoCommitStored`, so that every time you store the offset, both values are reset. Because right now is not when is triggered in the case `<-time.After(consumer.options.autoCommitStrategy.flushInterval)`.

**Comment by Gsantomaggio** [comment]
Thanks both. 
@hiimjako feel free to merge when it is ok with you

**Comment by Darthmineboy** [comment]
I'll make some improvements after work

File: pkg/stream/consumer_test.go (hunk 1 above)
Comment (Darthmineboy) [reply pkg/stream/consumer_test.go:312]: Consumer has to be closed no? The line above is closing the producer, so close() is not called twice on the consumer :)

**Comment by Darthmineboy** [comment]
> Looks great to me!
> 
> I was also thinking:
> 
> 1. If the lock in `cacheStoreOffset` is needed, because it doesn't look like there are concurrent calls to the function.
> 2. If it makes sense to move the reset of `messageCountBeforeStorage` into `cacheStoreOffset`, along with `lastAutoCommitStored`, so that every time you store the offset, both values are reset. Because right now is not when is triggered in the case `<-time.After(consumer.options.autoCommitStrategy.flushInterval)`.

1. I thought so as well, but the tests were failing due to a data race. What happens concurrently though is closing the consumer, which would cause the data race. I've made an additional change to resolve another data race issue.
2. I've moved resetting messageCountBeforeStorage so that it applies to both cases.

File: pkg/stream/consumer_test.go (hunk 1 above)
Comment (hiimjako) [reply pkg/stream/consumer_test.go:312]: Oh, I thought it was a duplicate line, sorry :sweat_smile:

**Comment by hiimjako** [comment]
> What happens concurrently though is closing the consumer, which would cause the data race.

Yeah, right, it also makes sense having both values in a lock.
To avoid it, maybe we could make a channel to stop the consumer routine and fill it at the beginning of the close function, so you don't have concurrent calls, but I would say leave it for future PR.

That's fine for me, thanks for the contribution :smile:

**Review by hiimjako — APPROVED** [review]

**Comment by Gsantomaggio** [comment]
Thanks @Darthmineboy @hiimjako

**Pull request merged**